### PR TITLE
Milestone 7: Validations, Model Specs, and N+1 problems

### PR DESCRIPTION
**Pull request by leopiresgoss**
## Milestone 7: Validations, Model Specs, and N+1 problems

In this milestone, I:

- Added  the following validations:
  - For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero
  - For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero
    - LikesCounter must be an integer greater than or equal to zero.
- Added unit specs for all of the models' methods and validations.
- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user

(Observation: I have already added the flash messages in the last milestone)

### CODE REVIEW

File: spec/models/comment_spec.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  it 'does not validate without an author' do
+    subject.post = nil
+    expect(subject).to_not be_valid
+  end
+
+  it 'does not validate without a text' do
+    subject.text = nil
+    expect(subject).to_not be_valid
+  end
+end
```
Comment (ahmetbozaci) [inline spec/models/comment_spec.rb]: - [x] Please fix the error.

image

File: spec/models/like_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  it 'does not validate without an author' do
+    subject.author = nil
+    expect(subject).to_not be_valid
+  end
+
+  it 'does not validate without an author' do
+    subject.post = nil
+    expect(subject).to_not be_valid
+  end
+end
```
Comment (ahmetbozaci) [inline spec/models/like_spec.rb]: - [x] Please fix the error.
image

File: spec/requests/posts_controller_spec.rb [hunk 3]
```diff
@@ -1,8 +1,13 @@
 require 'rails_helper'
 
 RSpec.describe '/posts', type: :request do
+  subject do
+    user = User.create(name: 'Tom', photo: '', bio: 'Teacher from Mexico.')
+    Post.create(author: user, title: 'Hello 2', text: 'This is my second post')
+  end
+
```
Comment (ahmetbozaci) [inline spec/requests/posts_controller_spec.rb:8]: - [x] Please fix the error.

image

**Review by ahmetbozaci — CHANGES_REQUESTED** [review]
# STATUS: Changes Required  ♻️


 @leopiresgoss  ✋🏼 

#### Your project is great. You are almost done. However, there are some issues you should fix. Please fix them then your project will be approved.

___ 

### 💯  Nice to have 💯
- Your git-flow is correct. ✔️
- You set up linter files and no errors. ✔️
- Your pull request is very professional. It has a meaningful title and summary. Keep it up.✔️
- You fix the n+1 problem in the post controller.
- You add the validations to the user and post the model according to the requirement.

___
### ♻️ Request changes ♻️

- [ ] _Check the comments under the review._
- [x] I wanted to check if ı can see any flash messages after creating posts, comments, and like. I click on the post and on the see all post button ı am getting an error. Please fix the error.
_Next code reviewer, please check the flash messages are working or not._
image
- [ ] You should also add tests for custom methods.(`increment_post_comments_counter, decrement_post_comments_counter, increment_user_posts_counter, five_recent_comments, decrement_user_posts_counter, increment_post_likes_counter, decrement_post_likes_counter,  tree_recent_posts,
update_comment_count`)  [Requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec_n%2B1.md#:~:text=all%20of%20your-,models%27%20methods,-and%20validations.)


___

### Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
___

### Cheers and Happy coding! 💻 👏
almost

___

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification._

_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._


___

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form.](https://airtable.com/shrQAqnBwek5a0O0s)_

[GitHub](https://github.com/ahmetbozaci) | [LinkedIn](https://linkedin.com/in/ahmetbozaci) | [Twitter](https://twitter.com/ahmtbozaci)

_Reviewed by [Ahmet Bozacı ](https://github.com/ahmetbozaci)_

File: spec/models/comment_spec.rb (hunk 1 above)
Comment (leopiresgoss) [reply spec/models/comment_spec.rb]: Thank you, I don't know why RSpec wasn't showing those errors

**Review by MahmoudBakr23 — APPROVED** [review]
Hi @leopiresgoss 😃, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! 🎉 

**Highlights**
- Well done making the required changes from the previous reviewer
- Great communication with the reviewer

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**